Conditionally Creating Server Spans if No Span Found in Current Context

andresbeckruiz opened the pull request:
# Description

This PR fixes issue 445 in the opentelemetry-python-contrib repo. I implemented a decorator class called ServerSpanTracer that checks if if an active span is present in the current context by calling `opentelemetry.trace.get_current_span()`. If one is found, an internal span is created and the span found is used as the parent. If one is not found, a server span is created and the remote span context from the incoming request is used as a parent.

I used either this class or logic for all of the server instrumentations that create server spans: Falcon, Flask, Django, Pyramid, Tornado, WSGI, and ASGI. There are references to the issue that mention that FastAPI and Starlette also create server spans, but I could not find where this occurs in the opentelemetry-python-contrib repo. If someone could point out where they are created I can add the logic I used for the other instrumentations. 

## Type of change

Please delete options that are not relevant.

- [ ✅ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
     -This feature causes server spans to be only created conditionally, so at times existing functionality will not work as expected. 

# How Has This Been Tested?

In the `opentelemetry.util.tracer` module, I wrote my tests in `server_span_tracer_test.py`. I wrote a single test to ensure that when a span is not found in the current context, a server span is created instead of an internal span. I plan to also test that when a span is found in the current context, an internal span is created instead of a server span; however, I noticed that `trace.get_current_span()` kept on returning `INVALID_SPAN` no matter how I tried implementing this test case, so I have emitted it for now. The test can be run by calling `tox -e test-util-tracer`.

# Does This PR Require a Core Repo Change?

- [ ] Yes. - Link to PR: 
- [ ✅ ] No.

# Checklist:

See [contributing.md](https://github.com/open-telemetry/opentelemetry-python-contrib/blob/main/CONTRIBUTING.md) for styleguide, changelog guidelines, and more.

- [ ✅ ] Followed the style guidelines of this project
- [ ✅ ] Changelogs have been updated
- [ ✅ ] Unit tests have been added
- [ ✅ ] Documentation has been updated

eddyleelin commented on util/opentelemetry-util-tracer/src/opentelemetry/util/tracer/__init__.py:
```diff
@@ -0,0 +1,48 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from opentelemetry import trace
+from opentelemetry.trace.span import Span
+from typing import Optional 
+from opentelemetry.util import types
```
I didn't know this until I just searched it up, but it seems like these module imports might need to be formatted a specific way (according to [the PEP 8 Python style guide](https://www.python.org/dev/peps/pep-0008/#imports)). Something like this, with an empty line separating the groups? 😄 
```suggestion
from typing import Optional

from opentelemetry import trace
from opentelemetry.trace.span import Span
from opentelemetry.util import types
```

eddyleelin commented on util/opentelemetry-util-tracer/src/opentelemetry/util/tracer/__init__.py:
```diff
@@ -0,0 +1,48 @@
+from opentelemetry.util import types
+
+class ServerSpanTracer:
+    "Decorator class for tracer for conditionally creating server spans"
```
Also wasn't sure about this until now, but it seems like this should be surrounded by triple quotes, even though it's still on one line 👍 [Check out PEP 257, under One-liner Docstrings](https://www.python.org/dev/peps/pep-0257/#id16). See for example [this](https://github.com/open-telemetry/opentelemetry-python-contrib/blob/9c834f088150e208d4fbd189477ed3bca5d106a3/opentelemetry-instrumentation/src/opentelemetry/instrumentation/propagators.py#L100)?
```suggestion
    """Decorator class for tracer for conditionally creating server spans."""
```

eddyleelin commented on util/opentelemetry-util-tracer/src/opentelemetry/util/tracer/__init__.py:
```diff
@@ -0,0 +1,48 @@
+        self.attributes = attributes
+
+    def get_span(self):
+        "Returns a new span with either the current span found as parent or remote span as parent"
```
According to PEP 257 (docstring conventions) that I linked above, it should technically be "return" rather than "returns", but it seems like they have written "returns" in other parts of the code... so I think it would be fine by just adding the period/triple quotes!
```suggestion
        """Returns a new span with either the current span found as parent or remote span as parent."""
```

eddyleelin commented on util/opentelemetry-util-tracer/tests/server_span_tracer_test.py:
```diff
@@ -0,0 +1,44 @@
+class ServerSpanTracerTest(TestCase):
+
+    def test_span_found_current_context(self):
+        "Tests that internal span is created when parent span found in the current context"
```
```suggestion
        """Tests that internal span is created when parent span is found in the current context."""
```

eddyleelin reviewed (commented):
Nice work!! This is really impressive and a lot of work you put in 😄 
Btw, do you know why the `util/opentelemetry-util-tracer/tests/__init__.py` file is empty?
Also, you might want to check to see if you can run the lint tests (I'm not sure why it's giving you the error right now, but will it run locally first maybe?

andresbeckruiz commented:
@eddyleelin thanks for the feedback! I just fixed lint issues in the code I added so the only issues left are from code that I have not touched. The reason `util/opentelemetry-util-tracer/tests/__init__.py` is empty is because all test folders in the repo have an `__init__.py` file (even if it's empty). I did not need to write any code in the file, but I wanted to make sure that I follow the proper package structure.

srprash commented on instrumentation/opentelemetry-instrumentation-asgi/tests/test_asgi_middleware.py:
```diff
@@ -135,7 +135,7 @@ def validate_outputs(self, outputs, error=None, modifiers=None):
             {
                 "name": "GET asgi",
-                "kind": trace_api.SpanKind.SERVER,
+                "kind": trace_api.SpanKind.INTERNAL,
```
Shouldn't this ASGI span be of kind SERVER? Is there another span wrapping it in the context?

andresbeckruiz replied:
Yes it should be, thank you

srprash commented on instrumentation/opentelemetry-instrumentation-django/tests/test_middleware.py:
```diff
@@ -206,7 +206,7 @@ def test_error(self):
             span.name, "^error/" if DJANGO_2_2 else "tests.views.error"
         )
-        self.assertEqual(span.kind, SpanKind.SERVER)
+        self.assertEqual(span.kind, SpanKind.INTERNAL)
```
You should test for SERVER spans as well.

srprash commented on util/opentelemetry-util-tracer/setup.cfg:
```diff
@@ -0,0 +1,50 @@
+    Programming Language :: Python :: 3
+    Programming Language :: Python :: 3.6
+    Programming Language :: Python :: 3.7
+    Programming Language :: Python :: 3.8
```
3.9 as well right?

srprash commented on util/opentelemetry-util-tracer/tests/server_span_tracer_test.py:
```diff
@@ -0,0 +1,43 @@
+
+class ServerSpanTracerTest(TestCase):
+
+    def test_span_found_current_context(self):
```
Also test for SERVER span creation when there is no span in current context.

andresbeckruiz replied:
I agree. I actually edited my start_span function because I realized that `INVALID_SPAN` would be returned instead of `None` for `trace.get_current_span()`, so now I was able to test for SERVER span creation. However, I am now unsure how to test for INTERNAL span creation, since `trace.get_current_span()`keeps on returning an `INVALID_SPAN ` for everything I have tried. I am going to ask upstream how to fix this problem so that I can test for INTERNAL span creation

srprash commented on util/opentelemetry-util-tracer/src/opentelemetry/util/tracer/__init__.py:
```diff
@@ -0,0 +1,48 @@
+        self.start_time = start_time
+        self.attributes = attributes
+
+    def get_span(self):
```
I think it should be renamed to `start_span` since it creates a new span and not just "get" an existing span,

srprash commented on instrumentation/opentelemetry-instrumentation-wsgi/tests/test_wsgi_middleware.py:
```diff
@@ -115,7 +115,7 @@ def validate_response(
         self.assertEqual(len(span_list), 1)
         self.assertEqual(span_list[0].name, span_name)
-        self.assertEqual(span_list[0].kind, trace_api.SpanKind.SERVER)
+        self.assertEqual(span_list[0].kind, trace_api.SpanKind.INTERNAL)
```
Why would this span be INTERNAL? Shouldn't it be SERVER kind?

srprash commented on util/opentelemetry-util-tracer/src/opentelemetry/util/tracer/__init__.py:
```diff
@@ -0,0 +1,48 @@
+    def get_span(self):
+        """Returns a new span with either the current span found as parent or remote span as parent."""
+        span = trace.get_current_span()
+        if span is None or not isinstance(span, Span):
```
When will the current span, if present in the context will not be of type `Span`? Pardon my unfamiliarity with OTel Python implementation. Also, what are the possible types of a span present in context?

andresbeckruiz replied:
You are right, I'm not sure why I was checking for this in the first place. The documentation specifies that the return type will always be of type `Span`. Thanks for the catch!

srprash commented on instrumentation/opentelemetry-instrumentation-asgi/src/opentelemetry/instrumentation/asgi/__init__.py:
```diff
@@ -203,9 +204,13 @@ async def __call__(self, scope, receive, send):
         token = context.attach(extract(scope, getter=asgi_getter))
         span_name, additional_attributes = self.span_details_callback(scope)
 
+        span_kind = trace.SpanKind.INTERNAL
+        current_span = trace.get_current_span()
+        if current_span is None or not isinstance(current_span, Span):
+            span_kind = trace.SpanKind.SERVER
```
Why not use `ServerSpanTracer` here as well?

andresbeckruiz replied:
I didn't use `ServerSpanTracer` here because for the asgi instrumentation, the span is created within a `try` block, which `ServerSpanTracer` does not implement because all the other instrumentations don't use one.

srprash commented on instrumentation/opentelemetry-instrumentation-django/src/opentelemetry/instrumentation/django/middleware.py:
```diff
@@ -126,13 +127,15 @@ def process_request(self, request):
+        server_span_tracer = ServerSpanTracer(
+            self._tracer, 
+            self._get_span_name(request), 
+            extract(request_meta, getter=wsgi_getter),
```
Why only use `wsgi_getter` in all instrumentations? What if we are using asgi server? Would it make a difference?

andresbeckruiz replied:
I'm not sure why wsgi_getter is used in all instrumentations, but it was used before I made my PR so I continued to use it when extracting context.

srprash commented on instrumentation/opentelemetry-instrumentation-django/src/opentelemetry/instrumentation/django/middleware.py:
```diff
@@ -33,6 +33,7 @@
 from opentelemetry.semconv.trace import SpanAttributes
 from opentelemetry.trace import Span, SpanKind, use_span
```
You can remove unused imports.

Commit:
Fixed context extraction and activation logic.
